src/sinks/aws_s3/sink.rs

```diff
@@ -225,92 +222,101 @@ where
     }
 }
 
-fn build_request(key: String, batch: Vec<Event>, options: &S3RequestOptions) -> S3Request {
-    // Generate the filename for this batch, which involves a surprising amount
-    // of code.
-    let filename = {
-        /*
-        Since this is generic over the partitioner, for purposes of unit tests,
-        we can't get the compiler to let us define a conversion trait such that
-        we can get &Event from &P::Item, or I at least don't know how to
-        trivially do that.  I'm leaving this snippet here because it embodies
-        the prior TODO comment of using the timestamp of the last event in the
-        batch rather than the current time.
-
-        Now that I think of it... is that even right?  Do customers want logs
-        with timestamps in them related to the last event contained within, or
-        do they want timestamps that include when the file was generated and
-        dropped into the bucket?  My gut says "time when the log dropped" but
-        maybe not...
-
-        let last_event_ts = batch
-            .items()
-            .iter()
-            .last()
-            .and_then(|e| match e.into() {
-                // If the event has a field called timestamp, in RFC3339 format, use that.
-                Event::Log(le) => le
-                    .get(log_schema().timestamp_key())
-                    .cloned()
-                    .and_then(|ts| match ts {
-                        Value::Timestamp(ts) => Some(ts),
-                        Value::Bytes(buf) => std::str::from_utf8(&buf)
-                            .ok()
-                            .and_then(|s| DateTime::parse_from_rfc3339(s).ok())
-                            .map(|dt| dt.with_timezone(&Utc)),
-                        _ => None,
-                    }),
-                // TODO: We don't ship metrics to the S3, but if we did, would this be right? or is
-                // there an actual field we should be checking similar to above?
-                Event::Metric(_) => Some(Utc::now()),
-            })
-            .unwrap_or_else(|| Utc::now());
-        let formatted_ts = last_event_ts.format(&time_format);
-        */
-        let formatted_ts = Utc::now().format(options.filename_time_format.as_str());
-
-        if options.filename_append_uuid {
-            let uuid = Uuid::new_v4();
-            format!("{}-{}", formatted_ts, uuid.to_hyphenated())
-        } else {
-            formatted_ts.to_string()
-        }
-    };
+pub trait S3RequestBuilder {
+    fn build_request(&self, key: String, batch: Vec<Event>) -> S3Request;
+}
```

Comment:
I like where this is going. Consistent with my notion that common bits could be pulled upward to support two unrelated code sites this trait allows the two sites to implement different builders. That's good. It's not great that the trait isn't lifted into a common place in the codebase but that's fixable. I would also encourage you to avoid boxed dyn traits where possible. Here's a [good article](https://medium.com/digitalfrontiers/rust-dynamic-dispatching-deep-dive-236a5896e49b) that deep dives and hits most of my concerns. You can instead add generic parameters to structs -- if you aren't sure how to do that, let me know and I'll help -- which allows us to do static dispatch instead.